arch/rv32i/src/pmp.rs

```diff
@@ -105,53 +115,77 @@ impl From<mpu::Permissions> for TORUserPMPCFG {
+/// non-zero bits in the 10 most significant bits will be rejected.
+///
 /// By accepting this type, PMP implementations can rely on these requirements
-/// to be verified. Furthermore, they can use the
-/// [`NAPOTRegionSpec::napot_addr`] convenience method to retrieve an `pmpaddrX`
-/// CSR value encoding this region's address and length.
+/// to be verified. Furthermore, they can use the [`NAPOTRegionSpec::pmpaddr`]
+/// convenience method to retrieve an `pmpaddrX` CSR value encoding this
+/// region's address and length.
 #[derive(Copy, Clone, Debug)]
 pub struct NAPOTRegionSpec {
-    start: *const u8,
-    size: usize,
+    pmpaddr: usize,
 }
 
 impl NAPOTRegionSpec {
-    /// Construct a new [`NAPOTRegionSpec`]
+    /// Construct a new [`NAPOTRegionSpec`] from a pmpaddr CSR value.
+    ///
+    /// For an RV32 platform, every single integer in `[0; usize::MAX]` is a
+    /// valid `pmpaddrX` CSR for a region configured in NAPOT mode, and this
+    /// operation is thus effectively infallible.
+    ///
+    /// For RV64 platforms, this operation checks if the range would include any
+    /// address outside of the 56 bit physical address space and, in this case,
+    /// rejects the `pmpaddr` (tests whether any of the 10 most significant bits
+    /// are non-zero).
+    pub fn from_pmpaddr_csr(pmpaddr: usize) -> Option<Self> {
+        // On 64-bit platforms, the 10 most significant bits must be 0:
+        if core::mem::size_of::<usize>() == core::mem::size_of::<u64>() {
+            if (pmpaddr as u64) & !PMPADDR_RV64_MASK != 0 {
+                return None;
+            }
+        }
+
+        Some(NAPOTRegionSpec { pmpaddr })
+    }
+
+    /// Construct a new [`NAPOTRegionSpec`] from a start address and size.
     ///
     /// This method accepts a `start` address and a region length. It returns
     /// `Some(region)` when all constraints specified in the
     /// [`NAPOTRegionSpec`]'s documentation are satisfied, otherwise `None`.
-    pub fn new(start: *const u8, size: usize) -> Option<Self> {
+    pub fn from_start_size(start: *const u8, size: usize) -> Option<Self> {
         if !size.is_power_of_two() || (start as usize) % size != 0 || size < 8 {
-            None
-        } else {
-            Some(NAPOTRegionSpec { start, size })
+            return None;
         }
-    }
 
-    /// Retrieve the start address of this [`NAPOTRegionSpec`].
-    pub fn start(&self) -> *const u8 {
-        self.start
-    }
-
-    /// Retrieve the size of this [`NAPOTRegionSpec`].
-    pub fn size(&self) -> usize {
-        self.size
+        Self::from_pmpaddr_csr(
+            ((start as usize) + (size - 1).overflowing_shr(1).0)
+                .overflowing_shr(2)
+                .0,
+        )
     }
 
-    /// Retrieve the end address of this [`NAPOTRegionSpec`].
-    pub fn end(&self) -> *const u8 {
-        unsafe { self.start.add(self.size) }
+    /// Construct a new [`NAPOTRegionSpec`] from a start .
```

Comment:
This sentence isn't complete (I think it was intended to say "from a start and end address.".